File: app/src/main/java/com/salman/bitclock/data/database/AppDatabase.java
package com.salman.bitclock.data.database;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import com.salman.bitclock.data.models.Alarm;
import com.salman.bitclock.data.models.Timer;

@Database(entities = {Alarm.class, Timer.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    // Access to Alarm data
    public abstract AlarmDao alarmDao();
    
    // Access to Timer data
    public abstract TimerDao timerDao();

    private static volatile AppDatabase INSTANCE;

    // Singleton pattern for database instance
    public static AppDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            AppDatabase.class, "bitclock_database")
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/kotlin/com/salman/bitclock/data/database/AppDatabase.kt
package com.salman.bitclock.data.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.salman.bitclock.data.models.Alarm
import com.salman.bitclock.data.models.Timer
import com.salman.bitclock.data.models.WorldClock

@Database(entities = [Alarm::class, Timer::class, WorldClock::class], version = 2, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {

    abstract fun alarmDao(): AlarmDao
    abstract fun timerDao(): TimerDao
    abstract fun worldClockDao(): WorldClockDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "bitclock_database"
                )
                    .fallbackToDestructiveMigration()
                    .build()
                INSTANCE = instance
                instance
            }
        }
    }
}
